R/mlaps.R

```diff
@@ -0,0 +1,395 @@
+  3013826, # Glucose Random
+  3040151, # Glucose Random
+  3018251, # Glucose Random
+  3006140  # Bilirubin
+)
+
+
+#' @title
+#' Assign score to LAPS component
+#'
+#' @param x (`character`)\cr
+#' A vector of result values for a particular lab test.
+#'
+#' @param breaks (`numeric`)\cr
+#' A vector of intervals for a particular test which correspond to a different number of points.
+#'
+#' @param points (`numeric`)\cr
+#' A vector of scores for each interval in `breaks`.
+#'
+#' @return (`numeric`)\cr
+#' A vector of scores for each lab result value provided in `x`.
+#'
+#' @importFrom stringr str_sub
+#' @export
+#'
+#' @examples
+#' x <- c("0.3", "0.7")
+#' breaks <- c(0, 0.2, 0.4, 0.8, Inf)
+#' points <- c(1, 2, 3, 4)
+#'
+#' laps_assign_test(x, breaks, points)
+#'
+laps_assign_test <- function(x, breaks, points) {
+  # casting as character then back to numeric because default output for `cut` with `labels` is `factor`
+  pts <- as.character(
+    cut(
+      suppressWarnings(as.numeric(x)),
+      breaks,
+      labels = points,
+      include.lowest = TRUE,
+      right = FALSE
+    )
+  )
+
+  if (sum(stringr::str_sub(x, 1, 1) %in% c("<", ">")) > 0) {
+    lower_limit <- breaks[2]
+    upper_limit <- breaks[length(breaks) - 1]
+    lower_points <- points[1]
+    upper_points <- points[length(points)]
+    pts[is.na(pts) & startsWith(x, "<") & suppressWarnings(as.numeric(stringr::str_sub(x, 2))) <= lower_limit] <- lower_points
+    pts[is.na(pts) & startsWith(x, ">") & suppressWarnings(as.numeric(stringr::str_sub(x, 2))) >= upper_limit] <- upper_points
+  }
+
+  return(as.numeric(pts))
+}
+
+
+#' @title
+#' Loop mLAPS
+#'
+#' @description
+#' A wrapper around the `mlaps()` function which breaks down the calculation of mLAPS on a
+#' large number of encounters by hospital-year. This avoids memory issues that can be caused
+#' by loading large chunks of the lab table.
+#'
+#' @param db (`DBIConnection`)\cr
+#' RPostgres DB connection.
+#'
+#' @param cohort (`numeric`)\cr
+#' A numeric vector of `genc_ids` to restrict the calculation to.
+#'
+#' @param hours_after_admission (`numeric`)\cr
+#' Consider lab tests collected **up to** `hours_after_admission` hours after inpatient admission in the calculation.
+#' Default `hours_after_admission` is set to 0, where only lab tests collected at Emergency Department (before inpatient admission) is considered in mLAPS calculation.
+#' Since not all encounters are admitted through Emergency Department, depending on research question, it can be relevant to consider lab tests collected in early inpatient admission.
+#' Typically, `hours_after_admission` can be set to 24 to consider any lab tests collected at Emergency Department and 24 hours after inpatient admission.
+#'
+#' @param output_laps_components (`logical`)\cr
+#' Does not aggregate the score and instead outputs for each LAPS component (test), its contribution
+#' to the overall score.
+#'
+#' @return (`data.frame`)\cr
```

Comment:
Maybe add here that patients without entry in the lab table are not returned in the output, so need to be merged with `cohort` again and their `mLAPS` value will be `NA`. Also worth mentioning that lab availability per `genc_id` partially depends on `hours_after_admission` (e.g., `genc_id` may have an entry in the lab table, but not within specified time window`.
Maybe this is too much here, but I'm wondering if we could also add something like this: "Note that in certain scenarios, it might be appropriate to impute missing LAPS with 0. For example, when `hours_after_admission` is set to 24, but some `genc_ids` did not have any lab tests performed within 24h of admission, it might be reasonable to assume that no lab tests were indicated (i.e., LAPS value should be close to 0). However, this assumption can only be made for patients who were hospitalized during time periods and hospitals where lab data were in principle available. Additionally, this assumption may not hold true for all patient cohorts, so researchers should use caution when imputing missing LAPS.